XMonad/Hooks/DynamicIcons.hs

```diff
@@ -30,112 +33,138 @@ import qualified XMonad.StackSet as S
+dynamicIconsPP :: IconConfig -> PP -> X PP
+dynamicIconsPP ic pp = getWorkspaceIcons ic <&> \ren -> pp{ ppRename = ppRename pp >=> ren }
+
+-- | Returns a function for 'ppRename' that augments workspaces with icons
+-- according to the provided 'IconConfig'.
+getWorkspaceIcons :: IconConfig -> X (String -> WindowSpace -> String)
+getWorkspaceIcons IconConfig{..} = fmt <$> getWorkspaceIcons' iconConfigIcons
+  where
+    fmt icons s w = iconConfigFmt s (M.findWithDefault [] (S.tag w) icons)
+
+getWorkspaceIcons' :: Query [String] -> X (M.Map WorkspaceId [String])
+getWorkspaceIcons' q = do
+    ws <- gets (S.workspaces . windowset)
+    is <- for ws $ foldMap (runQuery q) . S.integrate' . S.stack
+    pure $ M.fromList (zip (map S.tag ws) is)
+
 
 -- | Datatype for expanded 'Icon' configurations
 data IconConfig = IconConfig
-    { iconConfigIcons :: IconSet -- ^ The 'IconSet' to use
-    , iconConfigStack :: [String] -> String -- ^ The function to manage stacking of 'Icon's
-    , iconConfigPP    :: PP -- ^ The 'PP' to alter
+    { iconConfigIcons :: Query [String]
+      -- ^ What icons to use for each window.
+    , iconConfigFmt :: WorkspaceId -> [String] -> String
+      -- ^ How to format the result, see 'iconsFmtReplace', 'iconsFmtAppend'.
     }
 
 instance Default IconConfig where
     def = IconConfig
         { iconConfigIcons = mempty
-        , iconConfigStack = wrap "[" "]" . unwords
-        , iconConfigPP = def
-        }
-
--- | This is the same as 'dynamicLogIconsWithPP' but it takes a 'IconConfig'.
--- This allows you to manually customise the 'Icon's the stacking function and also your `PP`
-dynamicLogIconsConvert :: IconConfig -> X PP
-dynamicLogIconsConvert iconConfig = do
-    ws <- gets (S.workspaces . windowset)
-    icons <- M.fromList . catMaybes <$> mapM (getIcons (iconConfigIcons iconConfig)) ws
-    pure $ (iconConfigPP iconConfig)
-        { ppCurrent = ppSection ppCurrent iconCurrent icons
-        , ppVisible = ppSection ppVisible iconVisible icons
-        , ppHidden = ppSection ppHidden iconHidden icons
-        , ppHiddenNoWindows =  ppSection ppHiddenNoWindows iconHiddenNoWindows icons
+        , iconConfigFmt = iconsFmtReplace (wrapUnwords "{" "}")
         }
-  where
-    ppSection pF f icons = pF (iconConfigPP iconConfig) . concatIcons f . iconLookup icons
-    iconLookup icons x = M.findWithDefault [baseIconSet x] x icons
-    concatIcons f y
-        | length y > 1 = iconConfigStack iconConfig $ map f y
-        | otherwise = concatMap f y
 
+-- | 'iconConfigFmt' that replaces the workspace name with icons, if any.
+--
+-- First parameter specifies how to concatenate multiple icons. Useful values
+-- include: 'concat', 'unwords', 'wrapUnwords'.
+--
+-- >>> iconsFmtReplace concat "1" []
+-- "1"
+--
+-- >>> iconsFmtReplace concat "1" ["A", "B"]
+-- "AB"
+--
+-- >>> iconsFmtReplace (wrapUnwords "{" "}") "1" ["A", "B"]
+-- "{A B}"
```

Comment:
With three examples it may be worth thinking about putting them below a subsubheading (like  `=== Example`), so people have to click to see them.  This may or may not be an anti-feature, I can't quite decide.

Reply:
When I build the docs locally, subsubheadings aren't collapsed. Where did you see this working as you describe?

Reply:
So apparently it's subsubsubheadings...  Very readable :)  You can see it in action in the docs for the [massiv](https://hackage.haskell.org/package/massiv-0.6.0.0/docs/Data-Massiv-Array.html#v:singleton) library

Reply:
Actually there's a special syntax to make any header collapsible: https://haskell-haddock.readthedocs.io/en/latest/markup.html#headings

> As of 2.15.1, there’s experimental (read: subject to change or get removed) support for collapsible headers: simply wrap your existing header title in underscores, as per bold syntax. The collapsible section will stretch until the end of the comment or until a header of equal or smaller number of =s.

Reply:
But yeah, I think in this case it's nice. It was weird that these three helpers take more space in the docs than the rest of this module. :-)

![image](https://user-images.githubusercontent.com/300342/112327852-1178d480-8cb6-11eb-9039-c6992db374da.png)